src/Orbit/Graphics/Shader/Generator/Variables/IVariable.cpp

```diff
@@ -0,0 +1,159 @@
+#include "Orbit/Graphics/Shader/Generator/Variables/Vec3.h"
+#include "Orbit/Graphics/Shader/Generator/Variables/Vec4.h"
+#include "Orbit/Graphics/Shader/Generator/IGenerator.h"
+#include "Orbit/Graphics/Shader/Generator/MainFunction.h"
+#include "Orbit/Graphics/Shader/Generator/Swizzle.h"
+
+#include <map>
+#include <sstream>
+#include <string>
+
+ORB_NAMESPACE_BEGIN
+
+namespace ShaderGen
+{
+	IVariable::IVariable( const IVariable& other )
+		: m_value( other.m_value )
+		, m_type ( other.m_type )
+	{
+		other.SetUsed();
+	}
+
+	IVariable::IVariable( IVariable&& other )
+		: m_value ( std::move( other.m_value ) )
+		, m_type  ( other.m_type )
+		, m_stored( other.m_stored )
+		, m_used  ( other.m_used )
+	{
+		other.m_type   = VariableType::Unknown;
+		other.m_stored = false;
+		other.m_used   = false;
+	}
+
+	IVariable::IVariable( double f )
+		: m_value ( std::to_string( f ) )
+		, m_type  ( VariableType::Float )
+		, m_stored( false )
+		, m_used  ( false )
+	{
+	}
+
+	IVariable::IVariable( std::string_view value, VariableType type )
+		: m_value( value )
+		, m_type ( type )
+	{
+	}
+
+	void IVariable::StoreValue( void )
+	{
+		if( !m_stored )
+		{
+			static uint32_t unique_index = 0;
+
+			std::ostringstream ss;
+			ss << "local_" << ( unique_index++ );
+
+			auto value = m_value;
+			m_value = ss.str();
+
+			const std::string typestring( VariableTypeToString( m_type ) );
+			IGenerator::GetCurrentMainFunction()->code.append( "\t" + typestring + " " + m_value + " = " + value + ";\n" );
+
+			m_stored = true;
+		}
+	}
+
+	IVariable IVariable::operator*( const IVariable& rhs ) const
+	{
+		SetUsed();
+		rhs.SetUsed();
+
+		switch( IGenerator::GetCurrentMainFunction()->shader_language )
+		{
+			case ShaderLanguage::HLSL:
+			{
+				return IVariable( "mul( " + rhs.GetValue() + ", " + GetValue() + " )", GetType() );
+			} break;
+
+			default:
+			{
+				return IVariable( "( " + GetValue() + " * " + rhs.GetValue() + " )", GetType() );
+			} break;
+		}
+	}
+
+	IVariable IVariable::operator+( const IVariable& rhs ) const
+	{
+		SetUsed();
+		rhs.SetUsed();
+
+		return IVariable( "( " + GetValue() + " + " + rhs.GetValue() + " )", m_type );
+	}
+
+	IVariable IVariable::operator-( void ) const
+	{
+		SetUsed();
+
+		return IVariable( "( -" + GetValue() + " )", m_type );
+	}
+
+	Swizzle* IVariable::operator->( void )
+	{
+		static Swizzle swizzle;
+
+		SetUsed();
+
+		Swizzle::latest_accessed_variable = this;
+		return &swizzle;
+	}
+
+	void IVariable::operator=( const IVariable& rhs )
+	{
+		rhs.SetUsed();
+
+		StoreValue();
+		IGenerator::GetCurrentMainFunction()->code.append( "\t" + GetValue() + " = " + rhs.GetValue() + ";\n" );
+	}
+
+	void IVariable::operator+=( const IVariable& rhs )
+	{
+		rhs.SetUsed();
+
+		StoreValue();
+		IGenerator::GetCurrentMainFunction()->code.append( "\t" + GetValue() + " += " + rhs.GetValue() + ";\n" );
+	}
+
+	void IVariable::operator*=( const IVariable& rhs )
+	{
+		rhs.SetUsed();
```

Comment:
If `m_used` is under protected visibility, we don't need a `SetUsed` function. Same goes for `m_stored` and `SetStored`